httpobs/scanner/analyzer/headers.py

```diff
@@ -38,76 +82,86 @@ def content_security_policy(reqs: dict, expectation='csp-implemented-with-no-uns
+        # ever disappeared, then bad things could happen that had previously been prevented.
+        csp = {}
+        for k in set(list(headers['http'].keys()) + list(headers['equiv'].keys())):
+            csp[k] = headers['http'].get(k, set()).union(headers['equiv'].get(k, set()))
+    else:
+        csp = headers['http'] or headers['equiv']
+
+    # Get the various directives we look at
+    object_src = csp.get('object-src') or csp.get('default-src') or {'*'}
+    script_src = csp.get('script-src') or csp.get('default-src') or {'*'}
+    style_src = csp.get('style-src') or csp.get('default-src') or {'*'}
+
+    # Remove 'unsafe-inline' if nonce or hash are used are in script-src
+    # See: https://github.com/mozilla/http-observatory/issues/88
+    if any(source.startswith(('\'sha256-', '\'sha384-', '\'sha512-', '\'nonce-'))
+           for source in script_src) and '\'unsafe-inline\'' in script_src:
+        script_src.remove('\'unsafe-inline\'')
+
+    # Now to make the piggies squeal
+
+    # No 'unsafe-inline' or data: in script-src
+    # Also don't allow overly broad schemes such as https: in either object-src or script-src
+    # Likewise, if you don't have object-src or script-src defined, then all sources are allowed
+    if (script_src.intersection(dangerously_broad + unsafe_inline) or
+       object_src.intersection(dangerously_broad)):
+        output['result'] = 'csp-implemented-with-unsafe-inline'
+
+    # If the site is https, it shouldn't allow any http: as a source (passive or mixed content)
+    elif urlparse(response.url).scheme == 'https' and [d for d in set.union(*csp.values()) if 'http:' in d]:
+        output['result'] = 'csp-implemented-with-insecure-scheme'
+
+    # Don't allow 'unsafe-eval' in script-src or style-src
+    elif script_src.union(style_src).intersection({'\'unsafe-eval\''}):
+        output['result'] = 'csp-implemented-with-unsafe-eval'
+
+    # Don't allow 'unsafe-inline', data:, or overly broad sources in style-src
+    elif style_src.intersection(dangerously_broad + unsafe_inline):
+        output['result'] = 'csp-implemented-with-unsafe-inline-in-style-src-only'
+
+    # Only if default-src is 'none' and 'none' alone, since additional uris override 'none'
+    elif csp.get('default-src') == {'\'none\''}:
+        output['result'] = 'csp-implemented-with-no-unsafe-default-src-none'
+    else:
+        output['result'] = 'csp-implemented-with-no-unsafe'
 
-        # TODO: allow a small bonus for upgrade-insecure-requests?
+    # Once we're done, convert every set() in csp to an array
+    csp = {k: list(v) for k, v in csp.items()}
 
-        # Code defensively on the size of the data
-        output['data'] = csp if len(str(csp)) < 32768 else {}
+    # TODO: allow a small bonus for upgrade-insecure-requests?
 
-    else:
-        output['result'] = 'csp-not-implemented'
+    # Code defensively on the size of the data
+    output['data'] = csp if len(str(csp)) < 32768 else {}
```

Comment:
Would it make more sense to trim the string rather than remove it if it's long? Either `csp[:32768]` or `textwrap.shorten` (in 3.4+).

Reply:
I specifically just drop it because I don't want a truncated string to mess up future database queries, when I'm looking for feature usage.